### ISSUE TITLE
Android 60fps video flickers

### ISSUE THREAD

**Issue by HardBits** [Unity-Technologies/arfoundation-samples#416]
Hello there,

We're having an issue using ARFoundation on Android 60fps where the camera flickers the camera feed regularly. I've recorded a short video of it happening using the SimpleAR scene with only 1 standard assets model and no custom shaders or packages or anything. It's not the regular framerate dropping and the camera feed slowing down. This looks like the camera frames are shifting the position for a frame. Tested on Pixel 2 XL and Pixel 3 XL.

https://www.dropbox.com/sh/qcghtjsu2caz9in/AADpjgowTDJbifCcrUExcMMva?dl=0

In the videos you can see the closet behind the model shifting a few pixels to the sides. It gets really regular sometimes, and its hard to pinpoint what's causing this. Reducing to 30fps is the only way we found to work around this problem, but it's a really downgrade for high end devices.

Any help with this issue is really appreciated.

Using:
- Unity 2019.3.0f6
- AR Foundation 3.1.0 preview 4
- ARSubsystems 3.1.0 preview 4
- ARCore XR Plugin 3.1.0 preview 4
- XR Legacy Input Helpers 2.0.6

**Comment by tdmowrer** [comment]
A couple of us took a look at the videos and can't tell what you are trying to point out exactly. How are you setting the FPS? That is, are you only setting the [Unity application framerate](https://docs.unity3d.com/ScriptReference/Application-targetFrameRate.html), or are you setting a supported FPS 60 [camera configuration](https://docs.unity3d.com/Packages/com.unity.xr.arsubsystems@3.1/api/UnityEngine.XR.ARSubsystems.XRCameraSubsystem.html#UnityEngine_XR_ARSubsystems_XRCameraSubsystem_GetConfigurations_Allocator_) in ARFoundation?

**Comment by patryk191129** [comment]
Hello,

The newest update of Play Services for AR causes image flickering - 1.15.200121046
I have also tested on older version of Play Services (1.14.191118106) and I didn't experienced any flickering.
Flickers appears when application detects and keeps tracking marker.

These flickers appears on:
- Unity 2019.3.0f6
- ARFoundation 3.0.1
- ARKit XR Plugin 3.0.1
- ARCore XR Plugin 3.0.1
- ARSubsystems 3.0.0

**Comment by HardBits** [comment]
> A couple of us took a look at the videos and can't tell what you are trying to point out exactly. How are you setting the FPS? That is, are you only setting the [Unity application framerate](https://docs.unity3d.com/ScriptReference/Application-targetFrameRate.html), or are you setting a supported FPS 60 [camera configuration](https://docs.unity3d.com/Packages/com.unity.xr.arsubsystems@3.1/api/UnityEngine.XR.ARSubsystems.XRCameraSubsystem.html#UnityEngine_XR_ARSubsystems_XRCameraSubsystem_GetConfigurations_Allocator_) in ARFoundation?

We are setting it through the Application.framerate in our app, but the same happens without setting it. In that demo project, i didn't set anything, just let it run with all ARFoundation core and sample scripts. Regarding setting the camera configuration, i'm not sure what you mean. The ARCameraManager is on the camera gameobject with settings to auto and ARSession set to match the framerate.

I've added the specific parts of the videos when the flicker happens in slow motion on the same folder (https://www.dropbox.com/sh/qcghtjsu2caz9in/AADpjgowTDJbifCcrUExcMMva?dl=0). Hopefully it is easier to see now, at 60fps it happens too fast and it's easy to miss.

The flicker is exactly as @patryk191129 has shown in the link using Play Services 1.15, but also happens not using the imagetracking feature. I will try a build with the previous Play Services version and see if it solves the problem (thanks for the tip!).

**Comment by HardBits** [comment]
Okay, tried running the build with Play Services for AR 1.14.191118106 instead of the actual 1.15 but the problem persists. There's not a very noticeable difference using one or another. Also tried version 1.12 and 1.13 but the app shows a popup requesting update for the latest version.

**Comment by tdmowrer** [comment]
I think there may be two different issues being discussed in this thread.

@HardBits: On most devices, ARCore updates at 30 Hz. Setting the Unity framerate to 60 Hz will cause Unity to run at a different rate than ARCore, so you may end up with a few frames which use the same camera image from ARCore, and then the next frame might jump by a couple frames. In general, you should not set the Unity framerate separately. If you want 60 Hz, you should select a 60 Hz camera configuration and let the ARSession match the frame rate. [This sample](https://github.com/Unity-Technologies/arfoundation-samples/blob/master/Assets/Scripts/CameraConfigController.cs) shows how to do that. Note that most Android devices do not support 60 Hz. See https://developers.google.com/ar/discover/supported-devices. The Unity Profiler can also show you whether the Unity and ARCore frames are synchronized.

@patryk191129 We'll investigate the difference between 1.14 and 1.15 Play Services for AR.

**Comment by tdmowrer** [comment]
@patryk191129 Could you
1. [Submit a bug](https://unity3d.com/unity/qa/bug-reporting) with a project that reproduces the problem, then post the case number here.
1. Try to reproduce a similar problem with the image tracking samples in [Google's plugin](https://github.com/google-ar/arcore-unity-sdk).

**Comment by patryk191129** [comment]
> @patryk191129 Could you
> 
> 1. [Submit a bug](https://unity3d.com/unity/qa/bug-reporting) with a project that reproduces the problem, then post the case number here.

https://fogbugz.unity3d.com/default.asp?1223635_dlep5u4v0n6tvfq1

> 2. Try to reproduce a similar problem with the image tracking samples in [Google's plugin](https://github.com/google-ar/arcore-unity-sdk).

No issues on standalone ARCore

**Comment by HardBits** [comment]
Hey @tdmowrer, sorry for taking so long to test this out.

So i tried using the camera config example you showed but the results are the same. Again, testing on Pixel 2 XL that supports the 60fps mode. All of the 60fps modes used resulted in eventual flickers with time passing. I made sure not to include any script that would force Application.framerate to something and recorded a video using all the modes available. It's a bit hard to spot but if you go frame by frame, you can still clearly see it happening, a camera frame that's the same for 2 frames, but the virtual objects move.

https://www.dropbox.com/s/qwfau1m6k52adx9/CameraConfig_60fps_flicker.mp4?dl=0

60hz 640x480 (see at 37-38s)
60hz 1280x720 (see at 59s)
60hz 1920x1080 (see at 1:08s)

On the 30hz modes, this is not noticeable.

**Comment by tdmowrer** [comment]
@patryk191129 You have deselected the "Match Frame Rate" option in your ARSession Component. This designed to alleviate the exact problem you reporting. Is there a reason you need to disable this?

@HardBits The problem you have described would also be caused by disabling the "Match Frame Rate" option. Matching the frame rate means Unity won't render the next frame until it has a new frame from ARCore to render. If you disable this, then you would easily see two of the same background images in a row. Is this the case for you?

**Comment by patryk191129** [comment]
@tdmowrer 
Hello, many thanks for reply.
I can also disable option "Match Frame Rate" but it doesn't entilery fix the problem with camera flickering. I have noticed that issue happens when application can't handle stable framerate (even when you locked framerate to 30fps and it drops barely to 29fps). I think the issue with camera is caused by overlapping images from previous frames from AR camera. It only happens on Android build which got installed Google Play Services for AR 1.15 (on 1.14 it works fine). Obviously these issues didn't happened on IOS ;)

I hope you will find solution :)

**Comment by HardBits** [comment]
@tdmowrer I rechecked just to be sure, and i've had the Match Frame Rate flag checked since the start of the project, so its been active during all tests.

**Comment by tdmowrer** [comment]
@HardBits 
I've been trying to reproduce your issue without much luck (Pixel 3 running @ 60 Hz). One thing you can try is to print out the Unity [frame count](https://docs.unity3d.com/ScriptReference/Time-frameCount.html) along with the [timestamp](https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@3.0/api/UnityEngine.XR.ARFoundation.ARCameraFrameEventArgs.html#UnityEngine_XR_ARFoundation_ARCameraFrameEventArgs_timestampNs) for the ARCore camera image.

If you get two identical camera textures two frames in a row, then their timestamps should be identical was well. This is a way to verify the issue using the SDK without stepping frame by frame to find it visually. Here's a script I used to test this:

```csharp
using System;
using System.Collections.Generic;
using System.Text;
using Unity.Collections;
using UnityEngine;
using UnityEngine.Profiling;
using UnityEngine.XR.ARFoundation;

[RequireComponent(typeof(ARCameraManager))]
public class CameraConfigPicker : MonoBehaviour
{
    void OnEnable()
    {
        GetComponent<ARCameraManager>().frameReceived += OnFrameReceived;
        Log("Log initialized");
    }

    long? m_LastTimestamp;

    long? m_CurrentTimestamp;

    List<string> m_Log = new List<string>();

    void Log(string msg)
    {
        m_Log.Add(msg);
    }

    void OnFrameReceived(ARCameraFrameEventArgs eventArgs)
    {
        Profiler.BeginSample("OnFrameReceived");
        m_CurrentTimestamp = eventArgs.timestampNs;

        if (m_LastTimestamp == m_CurrentTimestamp)
        {
            Profiler.BeginSample($"Timestamps are the same! {m_LastTimestamp} -> {m_CurrentTimestamp}");
            Log($"{Time.frameCount}: {m_LastTimestamp} == {m_CurrentTimestamp}");
            Profiler.EndSample();
        }

        m_LastTimestamp = m_CurrentTimestamp;
        Profiler.EndSample();
    }

    void OnGUI()
    {
        GUI.skin.label.fontSize = 50;
        GUI.skin.button.fontSize = 50;
        
        var subsystem = GetComponent<ARCameraManager>().subsystem;
        if (subsystem != null)
        {
            GUILayout.Label($"Camera config: {subsystem.currentConfiguration}");
            GUILayout.Label($"{Time.frameCount}: {m_CurrentTimestamp}");
            foreach (var config in subsystem.GetConfigurations(Allocator.Temp))
            {
                if (GUILayout.Button(config.ToString()))
                {
                    subsystem.currentConfiguration = config;
                }
            }
            
            var sb = new StringBuilder();
            for (int i = m_Log.Count - 1; i > Mathf.Max(m_Log.Count - 10, 0); i--)
            {
                sb.Append(m_Log[i]);
                sb.Append("\n");
            }
            GUILayout.Label(sb.ToString());
        }
    }
}
```

Also, it would be helpful if you [submitted a bug](https://unity3d.com/unity/qa/bug-reporting) with your project so I can make sure I am testing the same thing.

**Comment by HardBits** [comment]
Hey @tdmowrer 

I've tested the project with the script you provided on the Pixel 2 XL. The only times i got equal frames, was when switching the camera mode. But there was also times along the video you can still clearly see it's the same camera frame and it didn't trigger it as that. At least with this script you provided, its easier to pinpoint the frames where it's happening.

[Video](https://www.dropbox.com/s/0wrhygr82sg9636/arfoundation_CameraConfigPicker%20_60fps_flicker.mp4?dl=0) with CameraConfigPicker

[Slowmo ](https://www.dropbox.com/s/bwp5ta76d6p4v1j/arfoundation_CameraConfigPicker%20_60fps_flicker_slowmo1.mp4?dl=0) of it happening in frame 1586 -> 1587

Other frames where you can see this happening:
1606 -> 1607
1610 -> 1611
1617 -> 1618

I've noticed some frames are being skipped along the video, but it looks more like a performance issue. This "flicker" might be related to it but looks to me it is a different problem.

I have reported the bug with the project attached as you requested: "(Case 1227258) ARFoundation 60fps camera flickers". Hope it helps. Let me know if you need any more help with testing the issue. Cheers

**Comment by tdmowrer** [comment]
> The only times i got equal frames, was when switching the camera mode...But there was also times along the video you can still clearly see it's the same camera frame and it didn't trigger it as that.

That means ARCore is providing us with a new [ArFrame](https://developers.google.com/ar/reference/c/group/frame) each frame. If there are duplicate background textures, then it sounds like something ARCore is doing.

> I have reported the bug with the project attached as you requested: "(Case 1227258) ARFoundation 60fps camera flickers". 

Thank you. I ran the project you attached on a Pixel 3. I don't notice anything visually, and I don't see any duplicate timestamps. How are you recording the videos you've been posting?

One thing I did notice is that the frame rate seems pretty spiky:
![Screen Shot 2020-03-20 at 5 46 19 PM](https://user-images.githubusercontent.com/4934281/77215858-b5440900-6ad3-11ea-9552-aa0d516c809a.png)

The spiky frames are just ARCore taking a bit longer than usual, but that means you aren't getting a smooth 60 fps, and the video capture software you're using might end up recording double frames from time to time.
